File: packages/protocol/go/edit_apply_result_test.go
package protocol

import "testing"

func TestEditApplyResultValidate(t *testing.T) {
	t.Parallel()

	success := NewEditApplySuccess([]EditAction{{
		Kind: "replace_between_anchors",
		Path: "/tmp/file.ts",
		Anchor: Anchor{
			Before: "before",
			After:  "after",
		},
		NewText: "updated",
	}})
	if err := success.Validate(); err != nil {
		t.Fatalf("expected success to validate, got %v", err)
	}

	failure := NewEditApplyFailure(EditFailure{
		Code:    "validation_failed",
		Message: "bad edit",
	})
	if err := failure.Validate(); err != nil {
		t.Fatalf("expected failure to validate, got %v", err)
	}

	noop := NewEditApplyNoop("No change needed.")
	if err := noop.Validate(); err != nil {
		t.Fatalf("expected noop to validate, got %v", err)
	}
}

func TestEditApplyResultValidateRejectsInvalidCombinations(t *testing.T) {
	t.Parallel()

	invalid := []EditApplyResult{
		{Kind: "success"},
		{
			Kind: "success",
			Actions: []EditAction{},
			Failure: &EditFailure{Code: "validation_failed", Message: "bad"},
		},
		{
			Kind: "failure",
			Failure: &EditFailure{Code: "validation_failed", Message: "bad"},
			Actions: []EditAction{{Kind: "replace_between_anchors"}},
		},
		{Kind: "noop"},
		{Kind: "unknown"},
	}

	for i, candidate := range invalid {
		if err := candidate.Validate(); err == nil {
			t.Fatalf("expected invalid result %d to fail validation", i)
		}
	}
}

File: packages/protocol/typescript/edit-apply-result.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import { isEditApplyResult } from "./validators";

test("isEditApplyResult accepts explicit success shape", () => {
  const value = {
    kind: "success",
    actions: [
      {
        kind: "replace_between_anchors",
        path: "/tmp/file.ts",
        anchor: { before: "a", after: "b" },
        newText: "x",
      },
    ],
  };

  assert.equal(isEditApplyResult(value), true);
});

test("isEditApplyResult rejects mixed success/failure shape", () => {
  const value = {
    kind: "success",
    actions: [],
    failure: {
      code: "validation_failed",
      message: "bad",
    },
  };

  assert.equal(isEditApplyResult(value), false);
});

test("isEditApplyResult accepts explicit noop shape", () => {
  const value = {
    kind: "noop",
    reason: "No change required.",
  };

  assert.equal(isEditApplyResult(value), true);
});

test("isEditApplyResult rejects extra keys on success", () => {
  const value = {
    kind: "success",
    actions: [],
    reason: "unexpected",
  };

  assert.equal(isEditApplyResult(value), false);
});

test("isEditApplyResult rejects invalid failure code", () => {
  const value = {
    kind: "failure",
    failure: {
      code: "random_error_code",
      message: "bad",
    },
  };

  assert.equal(isEditApplyResult(value), false);
});

